contracts/Presale.sol

```diff
@@ -0,0 +1,289 @@
+            roundIndex: $currentRoundIndex,
+            asset: DAI,
+            amountAsset: amount,
+            amountUSD: amount,
+            account: _sender
+        });
+
+        uint256 _purchaseAmountAsset = _sync(_purchaseConfig);
+        IERC20(DAI).transfer($config.withdrawTo, _purchaseAmountAsset);
+    }
+
+    receive() external payable {
+        purchase();
+    }
+
+    function _sync(PurchaseConfig memory purchaseConfig) private returns (uint256) {
+        PresaleConfig memory _config = $config;
+        RoundConfig[] memory _rounds = $rounds;
+
+        require(block.timestamp >= _config.startDate, "RAISE_NOT_STARTED");
+        require(block.timestamp <= _config.endDate, "RAISE_ENDED");
+        require(
+            purchaseConfig.amountUSD >= _config.minDepositAmount || _config.minDepositAmount == 0, "MIN_DEPOSIT_AMOUNT"
+        );
+
+        uint256 _totalPurchaseAmountAsset;
+        uint256 _remainingUSD = purchaseConfig.amountUSD;
+        uint256 _userAllocationRemaining = _config.maxUserAllocation - $userTokensAllocated[purchaseConfig.account];
+
+        uint256 _roundAllocationRemaining;
+
+
+        for (uint256 i = purchaseConfig.roundIndex; i < _rounds.length; ++i) {
+            uint256 _currentlyAllocated = $roundAllocated[i];
+            _roundAllocationRemaining =
+                _currentlyAllocated < _rounds[i].tokensAllocated ? _rounds[i].tokensAllocated - _currentlyAllocated : 0;
+
+            if (_roundAllocationRemaining == 0) continue;
+
+            uint256 _tokensAllocated = _remainingUSD / _rounds[i].tokenPrice;
+
+            if (_tokensAllocated > _roundAllocationRemaining) {
+                _tokensAllocated = _roundAllocationRemaining;
+            }
+            if (_tokensAllocated > _userAllocationRemaining) {
+                _tokensAllocated = _userAllocationRemaining;
+            }
+
+            if (_tokensAllocated > 0) {
+                uint256 _tokensCostUSD = _tokensAllocated * _rounds[i].tokenPrice;
+
+                _remainingUSD -= _tokensCostUSD;
+                _roundAllocationRemaining -= _tokensAllocated;
+                _userAllocationRemaining -= _tokensAllocated;
+
+                uint256 _roundPurchaseAmountAsset =
+                    _tokensCostUSD * purchaseConfig.amountAsset / purchaseConfig.amountUSD;
+                _totalPurchaseAmountAsset += _roundPurchaseAmountAsset;
+
+                PurchaseConfig memory _roundPurchaseConfig = PurchaseConfig({
+                    roundIndex: i,
+                    asset: purchaseConfig.asset,
+                    amountAsset: _roundPurchaseAmountAsset,
+                    amountUSD: _tokensCostUSD,
+                    account: purchaseConfig.account
+                });
+
+                _deposit(_roundPurchaseConfig, _tokensAllocated);
+            }
+        }
+
+        $userTokensAllocated[purchaseConfig.account] = _config.maxUserAllocation - _userAllocationRemaining;
+
```

Comment:
added require check for at least 1 token allocated, and set $currentRoundIndex at end of for loop

i left the userTokensAllocation assigning as it is since its one storage write outside of for loop instead of multiple inside